File: xssfinder-core/src/test/java/org/xssfinder/runner/WebDriverXPathFinderTest.java
package org.xssfinder.runner;

import com.google.common.collect.ImmutableList;
import org.junit.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;

import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class WebDriverXPathFinderTest {
    @Test
    public void elementWithIdReturnsXPathSelector() {
        // given
        WebDriverXPathFinder finder = new WebDriverXPathFinder();
        WebElement element = mock(WebElement.class);
        when(element.getTagName()).thenReturn("DIV");
        when(element.getAttribute("id")).thenReturn("someId");

        // when
        String xpath = finder.getXPath(element);

        // then
        assertThat(xpath, is("//div[@id=\"someId\"]"));
    }

    @Test
    public void bodyReturnsXPathSelector() {
        // given
        WebDriverXPathFinder finder = new WebDriverXPathFinder();
        WebElement element = mock(WebElement.class);
        when(element.getTagName()).thenReturn("BODY");

        // when
        String xpath = finder.getXPath(element);

        // then
        assertThat(xpath, is("body"));
    }

    @Test
    public void nondescriptElementReturnsXPathRelativeToAncestors() {
        // given
        WebDriverXPathFinder finder = new WebDriverXPathFinder();
        WebElement element = mock(WebElement.class);
        when(element.getTagName()).thenReturn("DIV");
        WebElement parent = mock(WebElement.class);
        when(parent.getTagName()).thenReturn("DIV");
        when(parent.getAttribute("id")).thenReturn("parentId");
        when(element.findElement(By.xpath(".."))).thenReturn(parent);
        when(parent.findElements(By.xpath("./*"))).thenReturn(ImmutableList.of(element));

        // when
        String xpath = finder.getXPath(element);

        // then
        assertThat(xpath, is("//div[@id=\"parentId\"]/div[1]"));
    }

    @Test
    public void indexIsDerivedFromParentsChildrenOfSameTag() {
        // given
        WebDriverXPathFinder finder = new WebDriverXPathFinder();
        WebElement element = mock(WebElement.class);
        when(element.getTagName()).thenReturn("DIV");
        WebElement parent = mock(WebElement.class);
        when(parent.getTagName()).thenReturn("DIV");
        when(parent.getAttribute("id")).thenReturn("parentId");
        when(element.findElement(By.xpath(".."))).thenReturn(parent);
        WebElement priorSibling = mock(WebElement.class);
        when(priorSibling.getTagName()).thenReturn("DIV");
        WebElement priorIgnoredSibling = mock(WebElement.class);
        when(priorIgnoredSibling.getTagName()).thenReturn("P");
        WebElement postSibling = mock(WebElement.class);
        when(postSibling.getTagName()).thenReturn("DIV");
        when(parent.findElements(By.xpath("./*"))).thenReturn(ImmutableList.of(
                priorSibling, priorIgnoredSibling, element, postSibling));

        // when
        String xpath = finder.getXPath(element);

        // then
        assertThat(xpath, is("//div[@id=\"parentId\"]/div[2]"));

    }
}

File: xssfinder-php-executor/test/XssFinder/Runner/WebDriverXPathFinderTest.php
<?php

namespace XssFinder\Runner;

use WebDriverElement;

class WebDriverXPathFinderTest extends \PHPUnit_Framework_TestCase
{
    function testElementWithIdReturnsXPathSelector()
    {
        // given
        $finder = new WebDriverXPathFinder();
        $element = $this->_mockElement('DIV', 'someId');

        // when
        $xpath = $finder->getXPath($element);

        // then
        assertThat($xpath, is('//div[@id="someId"]'));
    }

    function testBodyReturnsXPathSelector()
    {
        // given
        $finder = new WebDriverXPathFinder();
        $element = $this->_mockElement('BODY');

        // when
        $xpath = $finder->getXPath($element);

        // then
        assertThat($xpath, is('body'));
    }

    function testNondescriptElementReturnsXPathRelativeToAncestors()
    {
        // given
        $finder = new WebDriverXPathFinder();
        $element = $this->_mockElement('DIV');
        $parent = $this->_mockElement('DIV', 'parentId');
        $this->_setChildren($parent, $element);

        // when
        $xpath = $finder->getXPath($element);

        // then
        assertThat($xpath, is('//div[@id="parentId"]/div[1]'));
    }

    function testElementIndexIsDerivedFromParentsChildOfSameTag()
    {
        // given
        $finder = new WebDriverXPathFinder();
        $parent = $this->_mockElement('DIV', 'parentId');
        $element = $this->_mockElement('DIV');
        $priorSibling = $this->_mockElement('DIV');
        $priorIgnoredSibling = $this->_mockElement('P');
        $laterSibling = $this->_mockElement('DIV');
        $this->_setChildren($parent, $priorSibling, $priorIgnoredSibling, $element, $laterSibling);

        // when
        $xpath = $finder->getXPath($element);

        // then
        assertThat($xpath, is('//div[@id="parentId"]/div[2]'));
    }

    /**
     * @param string $tag
     * @param string $id
     * @return WebDriverElement
     */
    private function _mockElement($tag, $id=null) {
        /** @var WebDriverElement $element */
        $element = mock('WebDriverElement');
        when($element->getTagName())->return($tag);
        if ($id) {
            when($element->getAttribute('id'))->return($id);
        }
        return $element;
    }

    /**
     * @param WebDriverElement $parent
     * @param WebDriverElement $child1
     * @param WebDriverElement $anotherChild,...
     */
    private function _setChildren($parent, $child1) {
        $children = array();
        if (func_num_args()>1) {
            for ($i = 1; $i < func_num_args(); $i++) {
                $child = func_get_arg($i);
                $children[] = $child;
                when($child->findElement(\WebDriverBy::xpath("..")))->return($parent);
            }
        }
        when($parent->findElements(\WebDriverBy::xpath("./*")))->return($children);
    }
}